### PR TITLE
FIX accept both float and array-like for time_horizon in predict_proba

### PR DESCRIPTION
**Pull request by glemaitre**
There is a bug in `SurvivalBoost.predict_proba`. The `time_horizon` parameter does not currently accept array-like while it should.

### CODE REVIEW

File: hazardous/_survival_boost.py [hunk 1]
```diff
@@ -278,17 +280,19 @@ def fit(self, X, y, times=None):
...
-        time_horizon : float or list, default=None
+        time_horizon : int, float or array-like, default=None
+            The time horizon at which to estimate the probabilities. If `None`, the
+            `time_horizon` passed at the constructor is used. Therefore, this
+            parameter allows to override the `time_horizon` set at the constructor.
+
+        Returns
+        -------
+        incidence_probabilities : ndarray of shape (n_events + 1, n_samples, n_times)
+            The incidence probabilities for each event at the given time horizon.
```
Comment (ArturoAmorQ) [inline hazardous/_survival_boost.py:295]:
```suggestion
            The estimated probabilities at the given time horizon. The first entry
            at index 0 represents the survival probability (none of the events
            occurring), whereas the remaining entries correspond to the incidence
            probabilities for each specific event.
```

File: hazardous/_survival_boost.py [hunk 2]
```diff
@@ -278,17 +280,22 @@ def fit(self, X, y, times=None):
...
             The input samples.
-        time_horizon : float or list, default=None
+        time_horizon : int or float, default=None
+            The time horizon at which to estimate the probabilities. If `None`, the
+            `time_horizon` passed at the constructor is used. Therefore, this
+            parameter allows to override the `time_horizon` set at the constructor.
+
+        Returns
+        -------
+        y_proba : ndarray of shape (n_events + 1, n_samples)
```
Comment (ogrisel) [inline hazardous/_survival_boost.py:294]:
```suggestion
        y_proba : ndarray of shape (n_samples, n_events + 1)
```

File: hazardous/_survival_boost.py [hunk 3]
```diff
@@ -278,17 +280,22 @@ def fit(self, X, y, times=None):
...
+            The estimated probabilities at the given time horizon. The first entry
+            at index 0 represents the survival probability (none of the events
+            occurring), whereas the remaining entries correspond to the incidence
+            probabilities for each specific event.
```
Comment (ogrisel) [inline hazardous/_survival_boost.py:298]:
```suggestion
            The estimated probabilities at the given time horizon. The column
            indexed 0 stores the estimated probabilities of staying event-free at the
            requested time horizon for each observation described the matching row
            of X. The remaining columns store the estimated cumulated incidence (or 
            probability) for each event.
```

**Review by ogrisel — COMMENTED** [review]
Could you please fix the shape information of the `predict_cumulative_incidence` docstring? It lacks the time dimension. I believe it should be `(n_samples, n_events + 1, n_time_horizons)`.

Actually `test_survival_boost_incidence_and_survival` should probably have an assertion about the shape of the array returned by `predict_cumulative_incidence`.

File: hazardous/_survival_boost.py [hunk 4]
```diff
@@ -278,17 +280,22 @@ def fit(self, X, y, times=None):
...
-
         Parameters
         ----------
         X : array-like of shape (n_samples, n_features)
             The input samples.
-        time_horizon : float or list, default=None
+        time_horizon : int or float, default=None
+            The time horizon at which to estimate the probabilities. If `None`, the
+            `time_horizon` passed at the constructor is used. Therefore, this
+            parameter allows to override the `time_horizon` set at the constructor.
```
Comment (ogrisel) [inline hazardous/_survival_boost.py:290]:
```suggestion
            parameter allows to override the `time_horizon` passed to the constructor.
```

File: hazardous/_survival_boost.py [hunk 5]
```diff
@@ -278,17 +280,22 @@ def fit(self, X, y, times=None):
...
-        Returns a (n_events + 1)d array with shape (X.shape[0], n_events + 1).
-        The first column holds the survival probability to any event and others the
-        incicence probabilities for each event.
-
         Parameters
         ----------
         X : array-like of shape (n_samples, n_features)
             The input samples.
-        time_horizon : float or list, default=None
+        time_horizon : int or float, default=None
```
Comment (ogrisel) [inline hazardous/_survival_boost.py:287]:
```suggestion

        time_horizon : int or float, default=None
```

File: hazardous/tests/test_survival_boost.py [hunk 6]
```diff
@@ -66,7 +66,45 @@ def test_survival_boost_incidence_and_survival(seed):
...
+        est.predict_proba(X_test, time_horizon=[0, 1])
+
+    time_horizon = 0
+    y_pred = est.predict_proba(X_test, time_horizon=time_horizon)
+    assert y_pred.shape == (n_events, X_test.shape[0])
+    assert_allclose(y_pred.sum(axis=0), 1.0)
+
+    est.set_params(time_horizon=time_horizon)
+    y_pred = est.predict_proba(X_test)
+    assert y_pred.shape == (n_events, X_test.shape[0])
```
Comment (ogrisel) [inline hazardous/tests/test_survival_boost.py:106]: This should be:

```suggestion
    assert y_pred.shape == (X_test.shape[0], n_events)
```

to be consistent with scikit-learn classifiers.

File: hazardous/tests/test_survival_boost.py [hunk 7]
```diff
@@ -66,7 +66,45 @@ def test_survival_boost_incidence_and_survival(seed):
...
+    assert_allclose(y_pred.sum(axis=0), 1.0)
```
Comment (ogrisel) [inline hazardous/tests/test_survival_boost.py:107]:
```suggestion
    assert_allclose(y_pred.sum(axis=1), 1.0)
```

File: hazardous/tests/test_survival_boost.py [hunk 8]
```diff
@@ -66,7 +66,45 @@ def test_survival_boost_incidence_and_survival(seed):
...
```
Comment (ogrisel) [inline hazardous/tests/test_survival_boost.py:108]: As discussed IRL, we need to check that calling `.set_params(time_horizon=something)` followed by `.predict_proba(X_test)` is equivalent to calling `.predict_proba(X_test, time_horizon=something)`.

File: hazardous/_survival_boost.py [hunk 9]
```diff
@@ -278,17 +280,25 @@ def fit(self, X, y, times=None):
...
         ----------
         X : array-like of shape (n_samples, n_features)
             The input samples.
-        time_horizon : float or list, default=None
+
+        time_horizon : int or float, default=None
+            The time horizon at which to estimate the probabilities. If `None`, the
+            `time_horizon` passed at the constructor is used. Therefore, this
+            parameter allows to override the `time_horizon` parameter passed to
+            the constructor.
```
Comment (ogrisel) [inline hazardous/_survival_boost.py:292]: I think the last sentence is redundant.

```suggestion
            `time_horizon` passed at the constructor is used.
```

File: hazardous/_survival_boost.py [hunk 10]
```diff
@@ -278,17 +280,25 @@ def fit(self, X, y, times=None):
...
+
+        Returns
+        -------
+        y_proba : ndarray of shape (n_samples, n_events + 1)
+            The estimated probabilities at the given time horizon. The column
+            indexed 0 stores the estimated probabilities of staying event-free at the
+            requested time horizon for each observation described the matching row
+            of X. The remaining columns store the estimated cumulated incidence (or
+            probability) for each event.
```
Comment (ogrisel) [inline hazardous/_survival_boost.py:301]:
```suggestion
            indexed 0 stores the estimated probabilities of staying event-free at
            the requested time horizon for each observation described by the matching
            row of X. The remaining columns store the estimated cumulated incidence
            (or probability) for each event.
```

**Comment by glemaitre** [comment]
> Could you please fix the shape information of the predict_cumulative_incidence docstring? It lacks the time dimension. I believe it should be (n_samples, n_events + 1, n_time_horizons)

I would do that in a subsequent PR.

**Review by ogrisel — APPROVED** [review]

**Pull request merged**